keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -713,6 +717,74 @@ existing systems will be more continuous if we keep the meaning of
 `AssuredConcurrencyShares` has been renamed to
 `NominalConcurrencyShares`.
 
+In the midst of the `v1beta3` lifetime a field was added to
+`PriorityLevelConfigurationSpec` to make it possible to specify the
+`NominalConcurrencyShares` and `LendablePercent` of exempt priority
+levels.  That field is shown next.  Also, the definition of `sum_acs`
+in `LimitedPriorityLevelConfiguration` was updated to sum over all
+priority levels rather than just the non-exempt ones.  Before this
+change, the exempt priority levels did not get any nominal concurrency
+allocation nor lending limit and did not participate in borrowing ---
+they simply had unlimited dispatching and that had no relation with
+the dispatching for non-exempt priority levels.
+
+```go
+	// `exempt` specifies how requests are handled for an exempt priority level.
+	// This field MUST be empty if `type` is `"Limited"`.
+	// This field MAY be non-empty if `type` is `"Exempt"`.
+	// If empty and `type` is `"Exempt"` then the default values
+	// for `ExemptPriorityLevelConfiguration` apply.
+	// +optional
+	Exempt *ExemptPriorityLevelConfiguration
+```
+
+At the same time, the relevant new datatype was added.  It is shown below.
+
+```go
+type ExemptPriorityLevelConfiguration struct {
```

Comment:
I initially thought that we won't need any additional configuration for borrowing from exempt PL.

I understand the idea behind it (and reserving some capacity for exempt requests for all the time), but I'm wondering if it's not introducing additional cognitive complexity for users than it's worth it.

Given that you're not blocking dispatching of exempt requests anyway, there is still an ability of exceed capacity if number of exempt requests is growing over time.

So my personal preference would be to start simple and not introduce `ExemptPriorityLevelConfiguration` at all for now, and just  implicitly assume that:
- `NominalConcurrencyShares` for it is 0 [we start with 0, may grow over time due to borrowing]
- `LendablePercent` is 0

If it appears that we need more configuration, this can be added in the future, but I would actually prefer to start simple and wait for the request to have it configurable that comes from the real usecase.

WDYT?

@MikeSpreitzer 
@lavalamp @deads2k - for thoughts

Reply:
(1) I actually like giving the exempt requests some nominal concurrency.  That means they will not always borrow, so in some sense they are less intrusive on the rest of the system.

(2) I think it is good to make the configuration explicit.  Maybe if it is always zeroes then you could argue nothing needs to be explicit.  But as I mentioned above, I like non-zero values.

(3) Anyone wanting to understand what APF is going is going to have to see the complexity of what the exempt requests are doing.  Anyone who does not care can ignore it, even if the configuration for exempt requests is reified in API objects.

Reply:
The problem a cluster operator wants solved is that an exempt priority level should not overwhelm the server, for that it does not seem very intuitive for an admin to tweak the the fields of `ExemptPriorityLevelConfiguration` for an exempt PL.

can we try tackling this problem without adding any API fields? (I think @wojtek-t originally suggested the same). In the future, if need be, we can add new fields for exempt PL. 
For now, we can assign a default CL to each exempt PL and have it borrow from other non-exempt PLs without any limit.

Also, are we going to reject a request from an exempt PL if it can't borrow or the SCL has been reached? (this would not be in keeping with the current behavior)

Reply:
This change does not introduce limiting to the exempt priority levels, and the revised text attempts to make that clear.

If I understand correctly, based on our latest conversation, what concerned @wojtek-t the most was changing the concurrency limits for the non-exempt priority levels.  Taking some of the server's concurrency from them by default would do that.  Alternatively, dispatching could go over the server's concurrency limit by a certain amount by default --- but that would be confusing.

Another concern, not quite as strong, is the intrusiveness of the borrowing by the exempt priority level.  If it has a nominal concurrency of zero, then every adjustment period it will borrow a unique value and the behavior of all the other priority levels will be tweaked.  By allowing admins to configure a non-zero nominal concurrency for the exempt priority level, this makes it possible to not have the exempt priority level tweak all the non-exempt ones every time.  By allow a certain amount to be borrowable, that allows the admins a safety margin; it means that nominal concurrency limit for exempt does not have to be the one right number that is great for every adjustment period.